### PR TITLE
OSD-13240 Integration test for ONV

### PR DESCRIPTION
**Pull request by mjlshen**
## What does this PR do? / Related Issues / Jira
Adds an MVP integration test for [OSD-13240](https://issues.redhat.com//browse/OSD-13240)

This creates a BYOVPC with an AWS Network Firewall blocking quay.io following this architecture: 
https://docs.aws.amazon.com/network-firewall/latest/developerguide/arch-igw-ngw.html and runs osd-network-verifier's egress check before tearing down the BYOVPC. Future improvements are planned in the same story to add some concurrency as each full run takes ~20 minutes per region and there are 21 regions.

Why block quay.io?
* Part of the done criteria for this integration test was to validate that ONV does in fact run even if it can't pull down the image (i.e. it uses the image baked into the AMI).

## How to test this PR locally / Special Instructions
I added a README and the steps are there too, but essentially

```bash
go build .
./integration --region "${REGION}" --profile "${PROFILE}"
```

Here's an example against `af-south-1`

## Logs 
```
❯ go build . && ./integration --region af-south-1 --profile onv-admin
2022/09/27 14:01:12 using region af-south-1
2022/09/27 14:01:14 using AZ: af-south-1a
2022/09/27 14:01:14 created VPC: vpc-0faa8a183778f74a6
2022/09/27 14:01:15 enableDnsSupport is true for VPC: vpc-0faa8a183778f74a6
2022/09/27 14:01:15 enableDnsHostnames is true for VPC: vpc-0faa8a183778f74a6
2022/09/27 14:01:15 creating private subnet: 10.0.0.0/25
2022/09/27 14:01:16 waiting up to 1m0s for subnet to become available
2022/09/27 14:01:16 subnet available: subnet-0ba16b2baa182068f
2022/09/27 14:01:16 created private subnet: subnet-0ba16b2baa182068f
2022/09/27 14:01:16 creating firewall subnet: 10.0.0.128/25
2022/09/27 14:01:17 waiting up to 1m0s for subnet to become available
2022/09/27 14:01:17 subnet available: subnet-00163ae756e0afdcb
2022/09/27 14:01:18 created firewall subnet: subnet-00163ae756e0afdcb
2022/09/27 14:01:18 creating public subnet: 10.0.1.0/25
2022/09/27 14:01:18 waiting up to 1m0s for subnet to become available
2022/09/27 14:01:18 subnet available: subnet-0457007a03a0e3818
2022/09/27 14:01:19 created public subnet: subnet-0457007a03a0e3818
2022/09/27 14:01:20 created firewall rule group: arn:aws:network-firewall:af-south-1:429297027867:stateful-rulegroup/osd-network-verifier-rule-group-abcd
2022/09/27 14:01:21 created firewall policy: arn:aws:network-firewall:af-south-1:429297027867:firewall-policy/osd-network-verifier-firewall-policy-abcd
2022/09/27 14:01:22 created firewall: arn:aws:network-firewall:af-south-1:429297027867:firewall/osd-network-verifier-firewall-abcd
2022/09/27 14:01:22 allocated EIP address: eipalloc-0867246607be9d2ff
2022/09/27 14:01:23 created NAT gateway: nat-01a42c68fd2499568
2022/09/27 14:01:24 created internet gateway: igw-00df040a9f5f2484d
2022/09/27 14:01:25 created private subnet route table: rtb-0ab409390fa500936
2022/09/27 14:01:25 associated private subnet route table: rtbassoc-0b84502a32187ff80
2022/09/27 14:01:26 created public subnet route table: rtb-0ce054180d6f94ce6
2022/09/27 14:01:26 associated public subnet route table: rtbassoc-0cf8ca3a4b904b780
2022/09/27 14:01:27 created firewall subnet route table: rtb-033498ae9046c647e
2022/09/27 14:01:27 associated firewall subnet route table: rtbassoc-0bcabbbdacd42d9d6
2022/09/27 14:01:27 waiting up to 30s for internet gateway attachment to become available
2022/09/27 14:01:28 internet gateway attached: igw-00df040a9f5f2484d
2022/09/27 14:01:28 route created for firewall subnet 0.0.0.0/0 --> IGW
2022/09/27 14:01:28 created internet gateway route table: rtb-033498ae9046c647e
2022/09/27 14:01:29 associated internet gateway route table: rtbassoc-0b904bbe7fa60f8e6
2022/09/27 14:01:29 waiting up to 5m0s for the NAT Gateway to become available
2022/09/27 14:03:06 NAT gateway available: nat-01a42c68fd2499568
2022/09/27 14:03:06 route created for private subnet 0.0.0.0/0 --> NAT
2022/09/27 14:03:06 waiting up to 10m0s for the firewall to become READY
2022/09/27 14:03:07 status: PROVISIONING
2022/09/27 14:03:23 status: PROVISIONING
2022/09/27 14:03:38 status: PROVISIONING
2022/09/27 14:03:57 status: PROVISIONING
2022/09/27 14:04:48 status: PROVISIONING
2022/09/27 14:05:08 status: PROVISIONING
2022/09/27 14:06:03 status: READY
2022/09/27 14:06:03 firewall ready: arn:aws:network-firewall:af-south-1:429297027867:firewall/osd-network-verifier-firewall-abcd
2022/09/27 14:06:05 route created for public subnet 0.0.0.0/0 --> firewall
2022/09/27 14:06:06 Starting ONV egress validation
Created instance with ID: i-0e6ed3c2e89834405
EC2 Instance: i-0e6ed3c2e89834405 Running
Terminating ec2 instance with id i-0e6ed3c2e89834405
Summary:
printing out failures:
 -  egressURL error: Unable to reach quay.io:443
2022/09/27 14:09:00 ONV egress validation: Failure!
2022/09/27 14:09:02 deleting firewall arn:aws:network-firewall:af-south-1:429297027867:firewall/osd-network-verifier-firewall-abcd
2022/09/27 14:09:04 waiting up to 10m0s for the firewall to be deleted
2022/09/27 14:09:04 status: DELETING
2022/09/27 14:09:19 status: DELETING
2022/09/27 14:09:51 status: DELETING
2022/09/27 14:10:45 status: DELETING
2022/09/27 14:11:14 status: DELETING
2022/09/27 14:12:04 status: DELETING
2022/09/27 14:12:27 status: DELETING
2022/09/27 14:13:31 status: DELETING
2022/09/27 14:14:30 status: DELETING
2022/09/27 14:15:22 status: DELETING
2022/09/27 14:16:37 firewall deleted: arn:aws:network-firewall:af-south-1:429297027867:firewall/osd-network-verifier-firewall-abcd
2022/09/27 14:16:37 deleting firewall policy arn:aws:network-firewall:af-south-1:429297027867:firewall-policy/osd-network-verifier-firewall-policy-abcd
2022/09/27 14:16:38 waiting up to 10m0s for the firewall policy to be deleted
2022/09/27 14:16:38 status: DELETING
2022/09/27 14:16:54 firewall policy deleted: arn:aws:network-firewall:af-south-1:429297027867:firewall-policy/osd-network-verifier-firewall-policy-abcd
2022/09/27 14:16:54 deleting firewall rule group arn:aws:network-firewall:af-south-1:429297027867:stateful-rulegroup/osd-network-verifier-rule-group-abcd
2022/09/27 14:16:55 rule group deleted: arn:aws:network-firewall:af-south-1:429297027867:stateful-rulegroup/osd-network-verifier-rule-group-abcd
2022/09/27 14:16:55 deleting NAT Gateway: nat-01a42c68fd2499568
2022/09/27 14:16:56 waiting up to 5m0s for the NAT Gateway to be deleted
2022/09/27 14:18:34 NAT gateway deleted: nat-01a42c68fd2499568
2022/09/27 14:18:34 EIP released: eipalloc-0867246607be9d2ff
2022/09/27 14:18:35 detached internet gateway: igw-00df040a9f5f2484d
2022/09/27 14:18:35 internet gateway deleted: igw-00df040a9f5f2484d
2022/09/27 14:18:35 disassociated private subnet route table: rtbassoc-0b84502a32187ff80
2022/09/27 14:18:36 disassociated firewall subnet route table: rtbassoc-0bcabbbdacd42d9d6
2022/09/27 14:18:36 disassociated public subnet route table: rtbassoc-0cf8ca3a4b904b780
2022/09/27 14:18:36 disassociated internet gateway route table: rtbassoc-0b904bbe7fa60f8e6
2022/09/27 14:18:37 deleted private subnet route table: rtb-0ab409390fa500936
2022/09/27 14:18:37 deleted firewall subnet route table: rtb-033498ae9046c647e
2022/09/27 14:18:38 deleted public subnet route table: rtb-0ce054180d6f94ce6
2022/09/27 14:18:38 deleted internet gateway route table: rtb-0d2562c26ed832935
2022/09/27 14:18:38 deleted private subnet: subnet-0ba16b2baa182068f
2022/09/27 14:18:39 deleted firewall subnet: subnet-00163ae756e0afdcb
2022/09/27 14:18:39 deleted public subnet: subnet-0457007a03a0e3818
2022/09/27 14:18:40 deleted VPC: vpc-0faa8a183778f74a6
```

### CODE REVIEW

**Review by boranx — COMMENTED** [review]
nice suite!  I liked the way of handling the calls in NewIntegrationTestData() and its `Setup()` and `Cleanup()`
We might also include https://github.com/openshift/osd-network-verifier/blob/main/examples/aws/firewall/firewallConfig.go#L372-L380 to be testing failing scenarios
other than a small suggestion, all good from me

/lgtm

File: integration/main.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+	if err := data.Cleanup(context.TODO()); err != nil {
+		panic(err)
+	}
+}
+
+func onvEgressCheck(region, subnetId string) {
+	//---------Initialize required args---------
+	// Read AWS creds from environment
+	key, _ := os.LookupEnv("AWS_ACCESS_KEY_ID")
```
[the remark this answers is not in the document]
Comment (mjlshen) [reply integration/main.go:43]: Good ideas, will do!

File: integration/pkg/aws/aws.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2Types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/networkfirewall"
+	nfwTypes "github.com/aws/aws-sdk-go-v2/service/networkfirewall/types"
+	tags "github.com/aws/aws-sdk-go-v2/service/resourcegroupstaggingapi"
+)
+
+const (
+	// These are temporary because CIDR math is hard
```
Comment (fahlmant) [inline integration/pkg/aws/aws.go:16]: Using a `/24` for each subnet should be enough and easy math 😄

File: integration/pkg/aws/cleanup.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+	log.Printf("deleted VPC: %s", *id.vpcId)
+	id.vpcId = nil
+
+	return nil
+}
+
+// CleanupSubnets deletes the public/private subnets
+func (id *OnvIntegrationTestData) CleanupSubnets(ctx context.Context) error {
+	if id.publicSubnetId == nil || id.privateSubnetId == nil || id.firewallSubnetId == nil {
```
Comment (fahlmant) [inline integration/pkg/aws/cleanup.go:33]: Wouldn't some subnets get missed from deletion if one of them is blank?

File: integration/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjlshen) [reply integration/main.go:46]: I've implemented the firewall logic (block quay.io), still TODO to implement support for `--profile`

File: integration/pkg/aws/aws.go (hunk 2 above)
Comment (mjlshen) [reply integration/pkg/aws/aws.go:16]: At one point I was going to support a `--cidr` flag, but gave up after it wasn't super straightforward to do CIDR math in Go haha so I just made them consts. Good point that we can size this down to a `/24` though.

File: integration/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjlshen) [reply integration/main.go]: Just did this, so resolving this conversation!

**Comment by rafael-azevedo** [comment]
@mjlshen I have a PR coming with a big refactor in a few days by the end of this week is it okay if we hold this until that comes in

**Comment by mjlshen** [comment]
@rafael-azevedo I think this has no impact with respect to your refactor, most of the changes are in a subfolder `integration/` with only one small bugfix to a tagging function that I noticed via the test in `pkg/cloudclient/aws/private.go`

**Review by boranx — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**